### ISSUE TITLE
different losses used in generating target gradient and training input data

### ISSUE THREAD

**Issue by zliangak** [JonasGeiping/invertinggradients#2]
Hi Jonas, I am recently following your work and find it very interesting. When I am reproducing your experiment, I have one question. 

In line 105 of file "reconstruct_image.py",
https://github.com/JonasGeiping/invertinggradients/blob/414bfb93af379f29f19d5b94d3128990b4ee9dc7/reconstruct_image.py#L105
you use the loss defined in https://github.com/JonasGeiping/invertinggradients/blob/414bfb93af379f29f19d5b94d3128990b4ee9dc7/inversefed/data/loss.py#L85
where
https://github.com/JonasGeiping/invertinggradients/blob/414bfb93af379f29f19d5b94d3128990b4ee9dc7/inversefed/data/loss.py#L103
to generate the target gradient (input_gradient). I believe this loss is 0.5 * CrossEntropy.

However, when you try to reconstruct the input data, you use the loss
https://github.com/JonasGeiping/invertinggradients/blob/414bfb93af379f29f19d5b94d3128990b4ee9dc7/inversefed/reconstruction_algorithms.py#L59
in the GradientReconstructor class, which is directly CrossEntropy.

These two losses differ by 0.5. In other words, the loss used by the attacker is different from the loss of the data owner. Why is it? And I am wondering would it affects the result of the case "Eucl. Loss + L-BFGS" in Table 1 since this method is sensitive to gradient magnitude.

**Comment by JonasGeiping** [comment]
Ok, so I was finally able to come back to this issue.
I've included an updated script to run these baseline experiments for the full batch of 100 CIFAR-10 examples in the repo.
The results for table 1 in the main paper now look as follows:

-|LeNet | LeNet | ResNet 20-4 | ResNet 20-4
----|--------|-----| ------|-------
Trained Model: |False | True | False | True
L2+LBFGS| 33.11 | 7.84 | 18.61 | 2.44 
Prop.| 23.27 | 21.67 | 19.28 | 14.57


Some smaller comments:
 -  I've tried a range of hyperparameters for L2+LBFGS after the fix of the loss scaling, but the old settings are still optimal.
 - The fix most significantly improves the reconstruction quality for L2+LBFGS for the untrained model, otherwise however the same trends hold: For CIFAR10, the new algorithm is more robust and able to recover images even from trained models. The L-BFGS is only advantageous when the model is small and smooth, as it can converge faster if the problem is smooth enough. 
 - This table has a smaller value for the untrained LeNet for L-BFGS (barely within the old tolerance), this could be fixed by iterating even longer, as the solution is already very close to the true image, but I didn't have time to restart the computation. This would not help for the other cases, where the algorithm fails to recover something close to the true image (i.e. the algorithm finds a non-optimal local minimum of the recovery objective).
 - When a reconstruction diverges and returns NaNs (which can happen for L-BFGS), it is unclear what the PSNR score should be. This version consistently sets all NaNs to 0.0 psnr, which is the minimal value for an image, instead of skipping the image. (This effect is what reduces the values for L2+LBFGS for the trained LeNet)

I'd like to update this table on arxiv as well soonish, that should happen within the next month. Again, thank you for letting us know about this. Feel free to reopen this issue, if any questions remain.

Best, Jonas

**Comment by zliangak** [comment]
Hi, Jonas, many thanks for your new experiment and comprehensive explanation!
